sql/events.cc

```diff
@@ -1303,6 +1311,97 @@ Events::load_events_from_db(THD *thd)
   DBUG_RETURN(ret);
 }
 
+bool
+Events::search_n_execute_db_events(THD *thd,
+                                   Event_parse_data::enum_dbevent dbevent) {
+  DBUG_ENTER("Events::search_n_execute_db_events");
+  /* We need a temporary THD */
+  if (!thd)
+  {
+    if (!(thd= new THD(0)))
+    {
+      DBUG_RETURN(TRUE);
+    }
+    thd->thread_stack= (char*) &thd;
+    thd->store_globals();
+    thd->set_query_inner((char*) STRING_WITH_LEN("intern:Events::search_n_execute_db_events"),
+                         default_charset_info);
+    thd->set_time();
+  }
+
+  bool ret= TRUE;
+  TABLE *table;
+  READ_RECORD read_record_info;
+  DBUG_PRINT("enter", ("thd: %p", thd));
+
+  privilege_t saved_master_access(thd->security_ctx->master_access);
+  thd->security_ctx->master_access |= PRIV_IGNORE_READ_ONLY;
+  bool save_tx_read_only= thd->tx_read_only;
+  thd->tx_read_only= false;
+
+  ret= db_repository->open_event_table(thd, TL_WRITE, &table);
+
+  thd->tx_read_only= save_tx_read_only;
+  thd->security_ctx->master_access= saved_master_access;
+
+  if (ret)
+  {
+    my_message_sql(ER_STARTUP,
+                   "Event Scheduler: Failed to open table mysql.event",
+                   MYF(ME_ERROR_LOG));
+    DBUG_RETURN(TRUE);
+  }
+
+  if (init_read_record(&read_record_info, thd, table, NULL, NULL, 0, 1, FALSE))
+  {
+    close_thread_tables(thd);
+    DBUG_RETURN(TRUE);
+  }
```

Comment:
I will try to use it.

Reply:
Actually, I think it won't work. Events::load_events_from_db loads events into the event_queue, but I'm not using the event_queue for startup and shutdown triggers. I'm only loading SCHEDULE events into the event_queue to avoid any conflict in the event_scheduler code, since startup and shutdown triggers don't work based on time.

Should I make Events::load_events_from_db a two-in-one method that loads events into the event_queue or executes some generalized triggers?

Reply:
okay, if it won't make the code simpler — don't